crates/runtime/src/dataconnector.rs

```diff
@@ -391,48 +404,35 @@ impl<T: ListingTableConnector + Display> DataConnector for T {
             default_runtime_env(),
         );
 
-        let url = self
-            .get_object_store_url(dataset)
-            .context(InvalidConfigurationSnafu {
-                dataconnector: format!("{__self}"),
-                message: "Unable to parse URL",
-            })?;
+        let url = self.get_object_store_url(dataset)?;
 
-        let table_path =
-            ListingTableUrl::parse(url)
-                .boxed()
-                .context(InvalidConfigurationSnafu {
-                    dataconnector: format!("{__self}"),
-                    message: "Unable to parse URL",
-                })?;
-
-        let (file_format, extension) = self
-            .get_file_format_and_extension()
-            .map_err(Into::into)
-            .context(InvalidConfigurationSnafu {
-                dataconnector: format!("{__self}"),
-                message: "Unable to resolve file_format and file_extension",
-            })?;
+        // This shouldn't error because we've already validated the URL in `get_object_store_url`.
+        let table_path = ListingTableUrl::parse(url).boxed().context(InternalSnafu {
+            dataconnector: format!("{self}"),
+            code: "LTC-RP-LTUP".to_string(), // ListingTableConnector-ReadProvider-ListingTableUrlParse
+        })?;
+
+        let (file_format, extension) = self.get_file_format_and_extension()?;
         let options = ListingOptions::new(file_format).with_file_extension(&extension);
 
         let resolved_schema = options
             .infer_schema(&ctx.state(), &table_path)
             .await
             .boxed()
-            .context(InvalidConfigurationSnafu {
-                dataconnector: format!("{__self}"),
-                message: "Unable to infer files schema",
+            .context(UnableToConnectInternalSnafu {
+                dataconnector: format!("{self}"),
             })?;
```

Comment:
Is there a way we can inspect the error returned by `infer_schema` to see if its a username/password error? Then we could return a specific `DataConnectorError::UnableToConnectInvalidUsernameOrPassword`, instead of the more generic one.

Reply:
We can downcast Boxed errors to specific Error and than treat it accordingly

Reply:
> We can downcast Boxed errors to specific Error and than treat it accordingly
That should be done inside ftp connector, and mapped to `DataConnectorError`